### PR TITLE
Moves `@types/prettier` to this repository

### LINKED ISSUE #14033: RFC: Deprecates `@types/prettier` and maintains type definitions in `prettier/prettier` repository
## Background

- Prettier has JavaScript API.
- Type definitions for the API is maintained in [Definitely Typed](https://github.com/DefinitelyTyped/DefinitelyTyped).
  - https://www.npmjs.com/package/@types/prettier
  - https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/prettier

## Proposal

1. Moves `@types/prettier` to https://github.com/prettier/prettier.
2. Deprecates `@types/prettier`.

## Pros

- Definitely Typed is a huge repository, so it is slightly hard to fix for us.
- Easy to synchronize type definitions with Prettier's API implementation.
- API users do not need to install `@types/prettier`, need only `prettier`.
- Some plugins are implemented in TypeScript, so improving their development experience makes a lot of sense.
  - https://github.com/sveltejs/prettier-plugin-svelte/issues/317#issuecomment-1336552532

## Cons

- Increase the maintenance cost.
- Increase the npm package size.

## References

- https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html

### PR DESCRIPTION
**Pull request by sosukesuzuki**
## Description

Fixes #14033

This Pull Request moves type definitions for Prettier JS API that is in Definitely Typed to this repository. Please see the issue( #14033 ) for Pros/Cons and more details.

We provide type definitions for only the production release that is published via npm, and don't provide for GitHub direct install release.

~~Type definition files are put in `./scripts/build/dts-files/**/*.d.ts`.~~
~~The contents of the file are from the implementation in DefinitelyTyped, modified for the V3 changes.~~
~~The build script copies `./scripts/build/dts-files/**/*.d.ts` to `./dist`, and add `”types”: “./index.d.ts”` to the `package.json`.~~

Type definition files are put next to the implementation JS file that is an entry point for the build.
The contents of files are from DefinitelyTyped(modified for the V3 changes).
In build script, `buildJavascriptModule` function copies the alliance's type definition files next to itself, if necessary.

Unit tests for type definition files are put in `./tests/dts/unit`.
Tese cases are put in `./tests/dts/unit/cases/*.ts`.
`run.js` verify the TypeScript files contains in `./tests/dts/unit/cases/` by TypeScript Compiler API. If no errors occur for all files, the test will pass.
We use [ts-expect](https://github.com/TypeStrong/ts-expect) to verify type matching.

## References

- https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/prettier
- https://www.npmjs.com/package/@types/prettier
- [TypeScript: Documentation - Publishing](https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html)

## Checklist

- [x] I’ve added tests to confirm my change works.
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

File: scripts/build/build-package-json.js [hunk 1]
```diff
@@ -73,6 +73,7 @@ async function buildPackageJson({ files }) {
       "node -e \"assert.equal(require('.').version, require('..').version)\"",
   };
   packageJson.files = files.map(({ output: { file } }) => file).sort();
+  packageJson.types = "./index.d.ts";
```
Comment (fisker) [inline scripts/build/build-package-json.js:76]: Should we update `exports` field?

File: scripts/build/dts-files/doc.d.ts [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+  }
+
+  interface BreakParent {
+    type: "break-parent";
+  }
+
+  interface Concat {
+    type: "concat";
+    parts: Doc[];
+  }
```
Comment (fisker) [inline scripts/build/dts-files/doc.d.ts:36]: This already removed.

File: scripts/build/dts-files/doc.d.ts [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+    /**
+     * Specify the line length that the printer will wrap on.
+     * @default 80
+     */
+    printWidth: number;
+    /**
+     * Specify the number of spaces per indentation-level.
+     * @default 2
+     */
+    tabWidth: number;
```
Comment (fisker) [inline scripts/build/dts-files/doc.d.ts:196]: I don't think we have default value for `printWidth` and `tabWidth`.

File: scripts/build/dts-files/doc.d.ts [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+  function willBreak(doc: Doc): boolean;
+  function traverseDoc(
+    doc: Doc,
+    onEnter?: (doc: Doc) => void | boolean,
+    onExit?: (doc: Doc) => void,
+    shouldTraverseConditionalGroups?: boolean
+  ): void;
+  function findInDoc<T = Doc>(
+    doc: Doc,
+    callback: (doc: Doc) => T,
```
Comment (fisker) [inline scripts/build/dts-files/doc.d.ts:218]: Isn't it returns `boolean`?

File: scripts/build/dts-files/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,1091 @@
...
+  | "yaml";
+export type BuiltInParsers = Record<BuiltInParserName, BuiltInParser>;
+
+export type CustomParser = (
+  text: string,
+  options: Options
+) => AST | Promise<AST>;
+
+/**
+ * For use in `.prettierrc.js`, `.prettierrc.cjs`, `prettier.config.js` or `prettier.config.cjs`.
```
Comment (fisker) [inline scripts/build/dts-files/index.d.ts:293]: We support `.mjs` files too.

File: scripts/build/dts-files/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,1091 @@
...
+    /** @see [dedent](https://github.com/prettier/prettier/blob/main/commands.md#dedent) */
+    function dedent(doc: Doc): Align;
+    /** @see [dedentToRoot](https://github.com/prettier/prettier/blob/main/commands.md#dedenttoroot) */
+    function dedentToRoot(doc: Doc): Align;
+    /** @see [fill](https://github.com/prettier/prettier/blob/main/commands.md#fill) */
+    function fill(docs: Doc[]): Fill;
+    /** @see [group](https://github.com/prettier/prettier/blob/main/commands.md#group) */
+    function group(doc: Doc, opts?: GroupOptions): Group;
+    /** @see [hardline](https://github.com/prettier/prettier/blob/main/commands.md#hardline) */
+    const hardline: Concat;
```
Comment (fisker) [inline scripts/build/dts-files/index.d.ts:995]: typo?

**Comment by fisker** [comment]
Looks we are coping files from scripts to dist, why don't we put those files in `src` and next to the js file? Or even better, is it possible to define types in js file?

**Comment by sosukesuzuki** [comment]
> Looks we are coping files from scripts to dist, why don't we put those files in src and next to the js file?

Putting the type definition next to the implementation seems like a good idea.

> Or even better, is it possible to define types in js file?

I think our type definition is too complex to define in JS files.

File: scripts/build/build-package-json.js (hunk 1 above)
Comment (sosukesuzuki) [reply scripts/build/build-package-json.js:76]: I tried to use [TypeScript 4.7 exports](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-7.html#packagejson-exports-imports-and-self-referencing), but it did not work. At least in our current structure, this configuration works fine.

File: scripts/build/dts-files/doc.d.ts (hunk 4 above)
Comment (sosukesuzuki) [reply scripts/build/dts-files/doc.d.ts]: `findInDoc` returns not only `boolean`

**Comment by sosukesuzuki** [comment]
@fisker I've updated this PR from your review, please review.

File: scripts/build/config.mjs [hunk 7]
```diff
@@ -321,6 +321,7 @@ const pluginFiles = [
     umdVariableName,
     buildOptions,
     isPlugin: true,
+    withDts: true,
```
Comment (fisker) [inline scripts/build/config.mjs:324]: Why don't we check file exists instead of add this flag.

File: scripts/build/config.mjs [hunk 8]
```diff
@@ -387,6 +398,7 @@ const universalFiles = [...nonPluginUniversalFiles, ...pluginFiles].flatMap(
       platform: "universal",
       buildOptions,
       isPlugin,
+      withDts,
```
Comment (fisker) [inline scripts/build/config.mjs:401]: Aren't we copy every file twice?

Comment (sosukesuzuki) [reply scripts/build/config.mjs]: Oh thank you. I've addressed your review [a829a3b](https://github.com/prettier/prettier/pull/14212/commits/a829a3b56c16a872784f6c706dc1457af2de0583)

File: changelog_unreleased/api/14212.md [hunk 9]
```diff
@@ -0,0 +1,5 @@
+#### Moves `@types/prettier` to `prettier/prettier` repository (#14212 by @sosukesuzuki)
+
+Move the type definition files for TypeScript of the Prettier JavaScript API currently maintained in [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped/) to Prettier's repository.
+
+This saves users from having to install `@types/prettier`.
```
Comment (fisker) [inline changelog_unreleased/api/14212.md:5]: Can we just say we added `.d.ts` files.

File: scripts/build/build-package-json.js (hunk 1 above)
Comment (fisker) [reply scripts/build/build-package-json.js]: Can you try this?

[`1d60730`](https://github.com/prettier/prettier/pull/14212/commits/1d6073058b3c6225eb04d551bdd54c625be69063)

File: src/document/index.d.ts [hunk 10]
```diff
@@ -0,0 +1,8 @@
+import { doc } from "../index.js";
```
Comment (fisker) [inline src/document/index.d.ts:1]: Looks like we can't simply copy those files, but at least fix the path.

File: scripts/build/config.mjs [hunk 11]
```diff
@@ -16,6 +22,62 @@ const {
...
+  }
+  let data = await fs.promises.readFile(file.input, "utf8");
+  for (const { from, to } of replacements) {
+    data = data.replaceAll(
+      new RegExp(` from "${from}";`, "g"),
+      ` from "${to}";`
+    );
+  }
+  await writeFile(path.join(DIST_DIR, file.output.file), data);
+}
```
Comment (sosukesuzuki) [inline scripts/build/config.mjs:64]: @fisker I've added `typesFileBuilder`. It runs following steps:
- Checks if a target file includes the import path should be replaced to other path.
- If no, just copy file.
- If yes, replace import file path by `pathReplacementMap` and RegExp.

File: scripts/build/config.mjs [hunk 12]
```diff
@@ -16,6 +22,62 @@ const {
...
+    "src/language-js/parse/babel.d.ts": jsParsersImportReplacement,
+    "src/language-js/parse/flow.d.ts": jsParsersImportReplacement,
+    "src/language-js/parse/meriyah.d.ts": jsParsersImportReplacement,
+    "src/language-js/parse/typescript.d.ts": jsParsersImportReplacement,
+  };
+  const replacements = pathReplacementMap[file.input];
+  if (!replacements || replacements.length === 0) {
+    await copyFileBuilder({ file });
+    return;
+  }
```
Comment (fisker) [inline scripts/build/config.mjs:55]:
```suggestion
  const replacements = pathReplacementMap[file.input] ?? [];
```

File: scripts/build/config.mjs (hunk 11 above)
Comment (fisker) [reply scripts/build/config.mjs:64]: Actually we can calculate pathes (there is `files` in first argument can use to find where the entry build to), but we can improve this part later.

File: src/index.d.ts [hunk 13]
```diff
@@ -0,0 +1,1107 @@
...
+   */
+  singleQuote: boolean;
+  /**
+   * Use single quotes in JSX.
+   * @default false
+   */
+  jsxSingleQuote: boolean;
+  /**
+   * Print trailing commas wherever possible.
+   * @default 'all'
```
Comment (fisker) [inline src/index.d.ts:323]: Let's use double quote? We default to double quote, though we don't format JSDoc.

File: src/index.d.ts [hunk 14]
```diff
@@ -0,0 +1,1107 @@
...
+  | "Editor"
+  | "Format"
+  | "Other"
+  | "Output"
+  | "Global"
+  | "Special";
+
+export interface BaseSupportOption<Type extends SupportOptionType> {
+  readonly name?: string | undefined;
+  since: string;
```
Comment (fisker) [inline src/index.d.ts:650]: Let's remove `since`, we don't support it anymore.

File: src/index.d.ts [hunk 15]
```diff
@@ -0,0 +1,1107 @@
...
+
+export interface BooleanArraySupportOption
+  extends BaseSupportOption<"boolean"> {
+  default?: Array<{ value: boolean[] }> | undefined;
+  array: true;
+}
+
+export interface ChoiceSupportOption<Value = any>
+  extends BaseSupportOption<"choice"> {
+  default?: Value | Array<{ since: string; value: Value }> | undefined;
```
Comment (fisker) [inline src/index.d.ts:713]: No `since`.

File: src/index.d.ts [hunk 16]
```diff
@@ -0,0 +1,1107 @@
...
+ */
+export function getSupportInfo(): Promise<SupportInfo>;
+
+/**
+ * `version` field in `package.json`
+ */
+export const version: string;
+
+// https://github.com/prettier/prettier/blob/main/src/common/util-shared.js
+export namespace util {
```
Comment (fisker) [inline src/index.d.ts:778]: Can you change this part to the latest version in next branch?

File: src/index.d.ts [hunk 17]
```diff
@@ -0,0 +1,1107 @@
...
+    function indentIfBreak(
+      doc: Doc,
+      opts: { groupId: symbol; negate?: boolean | undefined }
+    ): IndentIfBreak;
+
+    /** @see [join](https://github.com/prettier/prettier/blob/main/commands.md#join) */
+    function join(sep: Doc, docs: Doc[]): Doc[];
+
+    /** @see [label](https://github.com/prettier/prettier/blob/main/commands.md#label) */
+    function label(label: string, doc: Doc): Label;
```
Comment (fisker) [inline src/index.d.ts:1021]: The first argument changed.

File: src/index.d.ts [hunk 18]
```diff
@@ -0,0 +1,1107 @@
...
+
+export interface ParserOptions<T = any> extends RequiredOptions {
+  locStart: (node: T) => number;
+  locEnd: (node: T) => number;
+  originalText: string;
+}
+
+export interface Plugin<T = any> {
+  languages?: SupportLanguage[] | undefined;
+  parsers?: { [parserName: string]: Parser<T> } | undefined;
```
Comment (fisker) [inline src/index.d.ts:434]: Parsers are allowed to be a function to lazy load(builtin parsers do this).

File: src/index.d.ts [hunk 19]
```diff
@@ -0,0 +1,1107 @@
...
+      doc: Doc,
+      callback: (doc: Doc) => T,
+      defaultValue: T
+    ): T;
+    function mapDoc<T = Doc>(doc: Doc, callback: (doc: Doc) => T): T;
+    function propagateBreaks(doc: Doc): void;
+    function removeLines(doc: Doc): Doc;
+    function stripTrailingHardline(doc: Doc): Doc;
+    function normalizeParts(parts: Doc[]): Doc[];
+    function normalizeDoc(doc: Doc): Doc;
```
Comment (fisker) [inline src/index.d.ts:1101]: These functions are not ready for plugins to use, should we remove from here?

File: tests/dts/unit/cases/parsers.ts [hunk 20]
```diff
@@ -0,0 +1,50 @@
+import * as prettier from "../../../../src/index";
+
+import acornEspreeParser from "../../../../src/language-js/parse/acorn-and-espree";
+import angularParser from "../../../../src/language-js/parse//angular";
+import babelParser from "../../../../src/language-js/parse//babel";
+import flowParser from "../../../../src/language-js/parse//flow";
```
Comment (fisker) [inline tests/dts/unit/cases/parsers.ts:6]: There are double slashes.

File: src/index.d.ts [hunk 21]
```diff
@@ -0,0 +1,1130 @@
...
+  get isFirst(): boolean;
+  get isLast(): boolean;
+  get isRoot(): boolean;
+  get root(): T;
+  get ancestors(): T[];
+
+  stack: T[];
+  callParent<U>(callback: (path: this) => U, count?: number): U;
+  getName(): PropertyKey | null;
+  getValue(): T;
```
Comment (fisker) [inline src/index.d.ts:99]: Can we deprecate these two methods in favor of `.key` / `.index` / `.node` getter?

Comment (fisker) [reply src/index.d.ts:99]: Where is `PropertyKey` ? Can't see.

File: src/index.d.ts [hunk 22]
```diff
@@ -0,0 +1,1130 @@
...
+  ): Doc;
+  embed?:
+    | ((
+        path: AstPath<T>,
+        print: (path: AstPath<T>) => Doc,
+        textToDoc: (text: string, options: Options) => Doc,
+        options: ParserOptions<T>
+      ) => Doc | null)
+    | undefined;
+  insertPragma?: ((text: string) => string) | undefined;
```
Comment (fisker) [inline src/index.d.ts:468]: Isn't it already marked optional?

File: src/index.d.ts [hunk 23]
```diff
@@ -0,0 +1,1130 @@
...
+    function indentIfBreak(
+      doc: Doc,
+      opts: { groupId: symbol; negate?: boolean | undefined }
+    ): IndentIfBreak;
+
+    /** @see [join](https://github.com/prettier/prettier/blob/main/commands.md#join) */
+    function join(sep: Doc, docs: Doc[]): Doc[];
+
+    /** @see [label](https://github.com/prettier/prettier/blob/main/commands.md#label) */
+    function label(label: any | undefined, contents: Doc): Label;
```
Comment (fisker) [inline src/index.d.ts:1046]: It not always returns `Label`, can be `contents`.

File: src/index.d.ts [hunk 24]
```diff
@@ -0,0 +1,1130 @@
...
+
+  function addLeadingComment(node: any, comment: any): void;
+
+  function addDanglingComment(node: any, comment: any, marker: any): void;
+
+  function addTrailingComment(node: any, comment: any): void;
+}
+
+// https://github.com/prettier/prettier/blob/next/src/document/index.js
+export namespace doc {
```
Comment (fisker) [inline src/index.d.ts:888]: Is it possible to move this to `src/document/index.js`? Since we bundle `doc.js` too.

File: src/index.d.ts (hunk 21 above)
Comment (sosukesuzuki) [reply src/index.d.ts:99]: `PropertyKey` is defined in TypeScript (`/node_modules/typescript`)

File: src/index.d.ts (hunk 22 above)
Comment (sosukesuzuki) [reply src/index.d.ts:468]: I think there was a compiler option that clarified the difference, but I forget. For now, I think it is `? ` is sufficient.

File: src/index.d.ts [hunk 25]
```diff
@@ -0,0 +1,909 @@
...
+  get root(): T;
+  get ancestors(): T[];
+
+  stack: T[];
+
+  callParent<U>(callback: (path: this) => U, count?: number): U;
+
+  /**
+   * @deprecated Please use `key`
+   */
```
Comment (fisker) [inline src/index.d.ts:108]:
```js
  /**
   * @deprecated Please use `AstPath#key` or  `AstPath#index`
   */
```

File: src/index.d.ts [hunk 26]
```diff
@@ -0,0 +1,909 @@
...
+  getName(): PropertyKey | null;
+
+  /**
+   * @deprecated Please use `node`
+   */
```
Comment (fisker) [inline src/index.d.ts:113]: /**
   * @deprecated Please use `AstPath#node` or  `AstPath#siblings`
   */

File: src/index.d.ts [hunk 27]
```diff
@@ -0,0 +1,909 @@
...
+  getValue(): T;
+
+  /**
+   * @deprecated Please use `node`
+   */
+  getNode(count?: number): T | null;
```
Comment (fisker) [inline src/index.d.ts:119]: This should not deprecate, it takes a argument, still useful.

File: src/index.d.ts [hunk 28]
```diff
@@ -0,0 +1,909 @@
...
+
+  /**
+   * @deprecated Please use `parent`
+   */
```
Comment (fisker) [inline src/index.d.ts:123]: Same here.

**Review by fisker — APPROVED** [review]
I don't have any experience on define types, but @type/prettier is there for years, I hope everything work as expected.

**Comment by kachkaev** [comment]
A PR that removes `@types/prettier` for v3: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/66073

**Comment by kachkaev** [comment]
`@types/prettier@3.0.0` has been released 🎉
https://www.npmjs.com/package/@types/prettier

Screenshot 2023-07-29 at 16 34 12


Thanks everyone who helped to move types into Prettier repo!

**Pull request merged**